File: python/deskcert/loader.py
"""Loads a task suite from either a directory (deskcert.config.yaml + tasks/*.yaml,
the layout `deskcert init` scaffolds) or a single YAML file holding the full
suite inline. Mirrors src/core/loader.ts.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any, Dict

import yaml

from .schema import validate_suite_object
from .types import SuiteConfig


def load_suite(suite_path: str) -> SuiteConfig:
    resolved = Path(suite_path).resolve()
    if not resolved.exists():
        raise FileNotFoundError(f'Suite path "{resolved}" does not exist')
    if resolved.is_dir():
        return _load_directory_suite(resolved)
    text = resolved.read_text(encoding="utf-8")
    return validate_suite_object(yaml.safe_load(text))


def _load_directory_suite(directory: Path) -> SuiteConfig:
    manifest_path = directory / "deskcert.config.yaml"
    if not manifest_path.exists():
        raise FileNotFoundError(f'No deskcert.config.yaml found in "{directory}"')
    manifest: Dict[str, Any] = yaml.safe_load(manifest_path.read_text(encoding="utf-8"))

    tasks_dir = directory / manifest.get("tasks_dir", "tasks")
    if not tasks_dir.exists():
        raise FileNotFoundError(f'Tasks directory "{tasks_dir}" does not exist')

    tasks = [
        yaml.safe_load(p.read_text(encoding="utf-8"))
        for p in sorted(tasks_dir.glob("*.yaml")) + sorted(tasks_dir.glob("*.yml"))
    ]

    suite_object: Dict[str, Any] = {"version": manifest["version"], "tasks": tasks}
    if manifest.get("pass_threshold") is not None:
        suite_object["pass_threshold"] = manifest["pass_threshold"]
    if manifest.get("forbidden_action_weight") is not None:
        suite_object["forbidden_action_weight"] = manifest["forbidden_action_weight"]

    return validate_suite_object(suite_object)

File: src/core/loader.ts
import { readFileSync, readdirSync, existsSync, statSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import { validateSuiteObject } from "./schema.js";
import type { SuiteConfig, TaskDefinition } from "./types.js";

interface SuiteManifest {
  version: number;
  pass_threshold?: number;
  forbidden_action_weight?: number;
  tasks_dir?: string;
}

/**
 * Loads a task suite from either:
 *  - a directory containing `deskcert.config.yaml` + a `tasks/` folder of
 *    one-task-per-file YAML definitions (the layout `deskcert init` scaffolds), or
 *  - a single YAML file holding the full suite (`version` + inline `tasks: [...]`).
 */
export function loadSuite(suitePath: string): SuiteConfig {
  const resolved = path.resolve(suitePath);
  if (!existsSync(resolved)) {
    throw new Error(`Suite path "${resolved}" does not exist`);
  }
  if (statSync(resolved).isDirectory()) {
    return loadDirectorySuite(resolved);
  }
  const text = readFileSync(resolved, "utf-8");
  return validateSuiteObject(yaml.load(text));
}

function loadDirectorySuite(dir: string): SuiteConfig {
  const manifestPath = path.join(dir, "deskcert.config.yaml");
  if (!existsSync(manifestPath)) {
    throw new Error(`No deskcert.config.yaml found in "${dir}"`);
  }
  const manifest = yaml.load(readFileSync(manifestPath, "utf-8")) as SuiteManifest;
  const tasksDir = path.join(dir, manifest.tasks_dir ?? "tasks");
  if (!existsSync(tasksDir)) {
    throw new Error(`Tasks directory "${tasksDir}" does not exist`);
  }
  const tasks: TaskDefinition[] = readdirSync(tasksDir)
    .filter((f) => f.endsWith(".yaml") || f.endsWith(".yml"))
    .sort()
    .map((f) => yaml.load(readFileSync(path.join(tasksDir, f), "utf-8")) as TaskDefinition);

  const suiteObject: Record<string, unknown> = { version: manifest.version, tasks };
  if (manifest.pass_threshold !== undefined) suiteObject.pass_threshold = manifest.pass_threshold;
  if (manifest.forbidden_action_weight !== undefined) {
    suiteObject.forbidden_action_weight = manifest.forbidden_action_weight;
  }
  return validateSuiteObject(suiteObject);
}
